### PR TITLE
Discovery tests for updating target groups

### PR DESCRIPTION
**Pull request by grandbora**
This pr
* Adds a test covering the case where a target providers sends updated versions of the same target groups and the system should reconcile to the latest version of each of the target groups
* Refactors how input data is represented in the tests. It used to be literal declarations of necessary structs, now it is parsing yaml. Yaml declarations are half as long as the former. And these can be put in a fixture file
* Adds a tiny bit of refactoring on test timeouts https://github.com/prometheus/prometheus/pull/3426/commits/10bad30ed799180ec65e1ea3c290bcf4d425e95f

For the second point, I think it is more developer friendly to read the yaml declarations from a fixture file. I'll work on it, drop a line if there are any opinions.

### CODE REVIEW

**Comment by grandbora** [comment]
cc @brian-brazil @fabxc please have a look.

File: discovery/testdata/single_provider_only_new_groups.yaml [hunk 1]
```diff
@@ -0,0 +1,195 @@
+# No updates.
```
Comment (brian-brazil) [inline discovery/testdata/single_provider_only_new_groups.yaml]: We generally put these inline with the test code.

Comment (grandbora) [reply discovery/testdata/single_provider_only_new_groups.yaml]: I pushed https://github.com/prometheus/prometheus/pull/3426/commits/21e9505d19b0b15bc4a4118f0a0f4f31741a1270 to address this

Comment (grandbora) [reply discovery/testdata/single_provider_only_new_groups.yaml]: This branch is green again. ping @brian-brazil

**Comment by grandbora** [comment]
ping @brian-brazil , anything else to address in this pr?

File: discovery/testdata/single_provider_only_new_groups.yaml (hunk 1 above)
Comment (brian-brazil) [reply discovery/testdata/single_provider_only_new_groups.yaml]: I meant that we don't have YAML files with Go objects, rather we have go objects inside the _test.go files.

Comment (grandbora) [reply discovery/testdata/single_provider_only_new_groups.yaml]: I started out with that approach but `_test.go` file became bloated with all the literal object declarations. You can see the example here : https://gist.githubusercontent.com/grandbora/bb4ef3af1588baa31e438054d438470f/raw/9d9d7eaca4bca739bed4e295d2e536e074a580ee/prometheus.tgs

It was really hard to work with that file afterwards. I can revert back to that but also if you have another suggestion I'd like to hear.

Comment (brian-brazil) [reply discovery/testdata/single_provider_only_new_groups.yaml]: This is what we do everywhere else, and we've far longer over in promql.

Comment (grandbora) [reply discovery/testdata/single_provider_only_new_groups.yaml]: Ok, I'll go back to that approach.

Comment (grandbora) [reply discovery/testdata/single_provider_only_new_groups.yaml]: Does this look good? https://github.com/prometheus/prometheus/pull/3426/commits/68344731087b9f7588ec17947f46a1f75a478a09

File: discovery/discovery_test.go [hunk 2]
```diff
@@ -278,9 +728,10 @@ func (tp mockTargetProvider) sendUpdates() {
 
 		time.Sleep(update.interval * time.Millisecond)
 
-		tgs := make([]*config.TargetGroup, len(update.targets))
-		for i, tg := range update.targets {
-			tgs[i] = &config.TargetGroup{Source: tg}
+		tgs := make([]*config.TargetGroup, len(update.targetGroups))
+		for groupIndex := range update.targetGroups {
+
```
Comment (grobie) [inline discovery/discovery_test.go]: Nit: remove empty line.

File: discovery/discovery_test.go [hunk 3]
```diff
@@ -278,9 +728,10 @@ func (tp mockTargetProvider) sendUpdates() {
 
 		time.Sleep(update.interval * time.Millisecond)
 
-		tgs := make([]*config.TargetGroup, len(update.targets))
-		for i, tg := range update.targets {
-			tgs[i] = &config.TargetGroup{Source: tg}
+		tgs := make([]*config.TargetGroup, len(update.targetGroups))
+		for groupIndex := range update.targetGroups {
```
Comment (grobie) [inline discovery/discovery_test.go]: Nit: `i` would be more in-line with the Go style guide, see https://github.com/golang/go/wiki/CodeReviewComments#variable-names.

File: discovery/discovery_test.go [hunk 4]
```diff
@@ -228,10 +667,21 @@ func TestTargetSetRunsSameTargetProviderMultipleTimes(t *testing.T) {
...
-	if *tp.callCount != 2 {
-		t.Errorf("Was expecting 2 calls received %v", tp.callCount)
+	finalize := make(chan struct{})
+	go func() {
+		defer close(finalize)
+		wg.Wait()
+	}()
+
+	select {
+	case <-time.After(20000 * time.Millisecond):
```
Comment (grobie) [inline discovery/discovery_test.go]: Nit: Why not `20 * time.Second`?

File: discovery/discovery_test.go [hunk 5]
```diff
@@ -228,10 +667,21 @@ func TestTargetSetRunsSameTargetProviderMultipleTimes(t *testing.T) {
...
+		t.Error("Test timed out after 20000 millisecond. All targets should be sent within the timeout")
+
+	case <-finalize:
+		if *tp.callCount != 2 {
+			t.Errorf("Was expecting 2 calls received %v", tp.callCount)
```
Comment (grobie) [inline discovery/discovery_test.go]: Nit: comma after "calls". Use `%d` for integers.

File: discovery/discovery_test.go [hunk 6]
```diff
@@ -15,136 +15,568 @@ package discovery
...
+		tp := newMockTargetProvider(testCase.updates)
+		targetProviders := map[string]TargetProvider{}
+		targetProviders["testProvider"] = tp
+
+		go targetSet.Run(ctx)
+		targetSet.UpdateProviders(targetProviders)
+
+		select {
+		case <-time.After(20000 * time.Millisecond):
+			t.Errorf("In test case %v[%v]: Test timed out after 20000 millisecond. All targets should be sent within the timeout", i, testCase.title)
```
Comment (grobie) [inline discovery/discovery_test.go]: Nit: use `%d` for integers, `%q` for strings. I'd shorten this to `%d. %q: timeout after 20 seconds. ...`

File: discovery/discovery_test.go [hunk 7]
```diff
@@ -15,136 +15,568 @@ package discovery
...
+						},
+					},
+					interval: 3000,
+				},
+			},
+		},
+	}
+
+	for i, testCase := range testCases {
+
```
Comment (grobie) [inline discovery/discovery_test.go]: Nit: remove empty line.

File: discovery/discovery_test.go [hunk 8]
```diff
@@ -15,136 +15,568 @@ package discovery
...
+					}
+				}
+
 			}
+		}
+	}
+}
+
+func TestTargetSetConsolidatesToTheLatestState(t *testing.T) {
+
```
Comment (grobie) [inline discovery/discovery_test.go]: Nit: remove emty line.

File: discovery/discovery_test.go [hunk 9]
```diff
@@ -15,136 +15,568 @@ package discovery
...
-
-		[]update{{[]string{"initial1", "initial2"}, 0}}, // Initials only.
-
-		[]update{{[]string{"initial1", "initial2"}, 6000}}, // Initials only but after a delay.
-
-		[]update{ // Initials and new groups.
-			{[]string{"initial1", "initial2"}, 0},
-			{[]string{"update1", "update2"}, 0},
+func TestTargetSetThrottlesTheSyncCalls(t *testing.T) {
+
```
Comment (grobie) [inline discovery/discovery_test.go]: Nit: remove empty line.

File: discovery/discovery_test.go [hunk 10]
```diff
@@ -15,136 +15,568 @@ package discovery
...
+					currentCallGroup[i] = tg.Source
 				}
+				syncedGroups = append(syncedGroups, currentCallGroup)
 
-				if len(tgs) == len(expectedGroups) {
+				syncCallCount++
+				if syncCallCount == len(testCase.expectedSyncCalls) {
 					// All the groups are sent, we can start asserting.
-					wg.Done()
+					finalize <- true
```
Comment (grobie) [inline discovery/discovery_test.go]: The idiomatic way in go is to just close the channel `close(finalize)`.

File: discovery/discovery_test.go [hunk 11]
```diff
@@ -15,136 +15,568 @@ package discovery
...
+					t.Errorf("In test case %v[%v]: received sync calls in call [%v]: \n %v \n do not match expected calls: \n %v \n", i, testCase.title, j, syncedGroups[j], testCase.expectedSyncCalls[j])
+				}
+
+				expectedGroupsMap := make(map[string]struct{})
+				for _, expectedGroup := range testCase.expectedSyncCalls[j] {
+					expectedGroupsMap[expectedGroup] = struct{}{}
 				}
+
+				for _, syncedGroup := range syncedGroups[j] {
+					if _, ok := expectedGroupsMap[syncedGroup]; ok == false {
```
Comment (grobie) [inline discovery/discovery_test.go]: Nit: It would be more idiomatic to use `!ok`.

File: discovery/discovery_test.go [hunk 12]
```diff
@@ -15,136 +15,568 @@ package discovery
...
+						t.Errorf("In test case %v[%v]: '%s' does not exist in expected target groups: %s", i, testCase.title, syncedGroup, testCase.expectedSyncCalls[j])
+					} else {
+						delete(expectedGroupsMap, syncedGroup) // Remove used targets from the map.
```
Comment (grobie) [inline discovery/discovery_test.go]: Why? The map will be cleaned up by the GC anyway.

File: discovery/discovery_test.go [hunk 13]
```diff
@@ -15,136 +15,568 @@ package discovery
...
+					interval:     10,
+				},
+			},
+			expectedSyncCalls: [][]string{
+				{"initial1"},
+				{"initial1", "update1", "update2", "update3"},
+			},
+		},
+		{
+			title: "Next test case",
```
Comment (grobie) [inline discovery/discovery_test.go]: Would be great to use unique and descriptive test titles instead.

File: discovery/discovery_test.go (hunk 12 above)
Comment (grandbora) [reply discovery/discovery_test.go]: This is not for cleaning up the the memory. It is for detecting the failures if the `sut` returns duplicate items. 
eg.
```
actualSyncedItems := []string{"tg1", "tg1"}
expectedSyncItems := map[string]struct{}{"tg1":... , "tg2":...}
```
Unless we delete `tg1` after it is used test would still pass.

**Comment by grandbora** [comment]
Thanks for the comments (the useful ones 😛 ) @grobie. I added https://github.com/prometheus/prometheus/pull/3426/commits/099094dff75b10a039b85e6b53d86d138ee36e64 to address them.

**Comment by grandbora** [comment]
I have couple of other additions in mind for these tests. It would be nice if this pr can be merged in before it gets too huge to review. ❤️

**Comment by grobie** [comment]
Thanks!
